woocommerce/woocommerce-ios
# Fix Crash When Tapping on Review Push Notification

Fixes #2254. 

~⚠️  Depends on #2668. Please review that first. 🙂~

## Findings

The [crash](https://sentry.io/organizations/a8c/issues/1487125043/?project=1458804&referrer=github_integration) happens in the `fatalError()` in this line:

https://github.com/woocommerce/woocommerce-ios/blob/061c2bce126fa84064d2125c45339be65269ab7a/WooCommerce/Classes/Tools/SyncCoordinator.swift#L123-L128

That points to the possibility that this provisioning method that sets up the `syncingCoordinator` was never called:

https://github.com/woocommerce/woocommerce-ios/blob/061c2bce126fa84064d2125c45339be65269ab7a/WooCommerce/Classes/ViewRelated/Notifications/ReviewsViewController.swift#L149-L153

This means that the caller of that method, [`ReviewsViewController.viewDidLoad`](https://github.com/woocommerce/woocommerce-ios/blob/061c2bce126fa84064d2125c45339be65269ab7a/WooCommerce/Classes/ViewRelated/Notifications/ReviewsViewController.swift#L109), was never called too. This seems impossible. And I failed to reproduce it. 😞  I guessed that might be a race condition. 

### Complex Flow

I also found that the way we handle a review push notification is complex and can be inefficient.

![](https://user-images.githubusercontent.com/198826/85342748-865d7a80-b4a8-11ea-8f9d-9b9e281fffdf.png)

There are at least 4 API calls fetching lots of objects. And, in the context of the push notification flow, we also rely on an unrelated `ReviewsViewController` to be _correctly synchronized_ in order to present the `ReviewDetailsViewController`.

## Solution

With all this information, I set on to fix this by:

- Optimizing the number of API calls that need to happen.
- Limit the dependency on a `UIViewController` to be properly set up. 

This is what the flow now looks like:

![](https://user-images.githubusercontent.com/198826/85342852-ba38a000-b4a8-11ea-910a-057ce811c44f.png)

1. Created a [`ReviewsCoordinator`](https://github.com/woocommerce/woocommerce-ios/blob/issue/2254-fix-review-push-notification-crash-2/WooCommerce/Classes/ViewRelated/Notifications/ReviewsCoordinator.swift) which is [a child of `MainTabBarController`](https://github.com/woocommerce/woocommerce-ios/blob/08758e67b7a27f631cdd20de0e5a6caa6c63db06/WooCommerce/Classes/ViewRelated/MainTabBarController.swift#L101-L103).
2. Made the `ReviewsCoordinator` [observe notifications from `PushNotesManager`](https://github.com/woocommerce/woocommerce-ios/blob/08758e67b7a27f631cdd20de0e5a6caa6c63db06/WooCommerce/Classes/ViewRelated/Notifications/ReviewsCoordinator.swift#L48-L52). This gives **full control** of how push notifications are handled to `ReviewsCoordinator` and should make unit testing easier. 
3. For every review push notification, dispatch a new Yosemite action, [`ProductReviewAction. retrieveProductReviewFromNote`](https://github.com/woocommerce/woocommerce-ios/blob/08758e67b7a27f631cdd20de0e5a6caa6c63db06/Yosemite/Yosemite/Actions/ProductReviewAction.swift#L21-L27) (added in #2458). This action fetches the `Note`, then the `ProductReview`, and then the `Product` and returns all of them at once. 
4. Present the `ReviewDetailsViewController` [from the `ReviewsCoordinator`](https://github.com/woocommerce/woocommerce-ios/blob/08758e67b7a27f631cdd20de0e5a6caa6c63db06/WooCommerce/Classes/ViewRelated/Notifications/ReviewsCoordinator.swift#L80). This **eliminates** the `ReviewsViewController` dependency. 

With this, whenever we present the `ReviewDetailsVC`, we are confident that we have all the objects that we need and we don't end up with the scenario shown by #1907. This should also (hopefully) eliminate the #2254 crash.

## Testing

1. Enable push notifications in your sandbox. 
2. Log in and navigate to a tab that is not Reviews. 
3. Place the app in the background.
4. Submit a product review on the web. 
5. Wait for the push notification. If it doesn't come, maybe check the Spam section in Comments. I found that they sometimes get sent there. Loggin in with a user and then submitting a review as that user seems to work better.
6. Tap on the push notification.
7. Wait for a second. 
8. Confirm that the Reviews tab was opened. 
9. Confirm that the Review Details page was presented. 
10. Confirm that the Review Details shows the Product name and the Review comment. 

## Testing (Auto-store switching)

1. Enable push notifications in your sandbox. 
2. Launch the app on your device, subscribe the device for a specific store (remind it).
3. Close the app.
3. Open the code, comment the two methods `unregisterForRemoteNotifications()` and `registerDeviceToken()` in `PushNotificationsManager`.
4. Re-run the app, and change the store.
4. Place the app in the background.
5. Submit a product review on the **previous store** that you started with.
6. Wait for the push notification. 
7. Tap on the push notification.
7. Wait for a second. 
8. Confirm that the Reviews tab was opened. 
9. Confirm that the Review Details page was presented. 
10. Confirm that the Review Details shows the Product name and the Review comment. 
11. Confirm that the current store was switched.


Please also test that you can still receive and tap on new Order push notifications.

## Submitter Checklist

- [ ] Mark Sentry as resolved.
- [ ] Create an issue for Order notifications
